Lib/fontTools/ttLib/tables/_g_l_y_f.py

```diff
@@ -243,6 +243,154 @@ def __len__(self):
 		assert len(self.glyphOrder) == len(self.glyphs)
 		return len(self.glyphs)
 
+	def getPhantomPoints(self, glyphName, ttFont, defaultVerticalOrigin=None):
+		"""Compute the four "phantom points" for the given glyph from its bounding box
+		and the horizontal and vertical advance widths and sidebearings stored in the
+		ttFont's "hmtx" and "vmtx" tables.
+
+		If the ttFont doesn't contain a "vmtx" table, the hhea.ascent is used as the
+		vertical origin, and the head.unitsPerEm as the vertical advance.
+
+		The "defaultVerticalOrigin" (Optional[int]) is used when the ttFont contains
+		neither a "vmtx" nor an "hhea" table.
+
+		https://docs.microsoft.com/en-us/typography/opentype/spec/tt_instructing_glyphs#phantoms
+		"""
+		glyph = self[glyphName]
+		assert glyphName in ttFont["hmtx"].metrics, ttFont["hmtx"].metrics
+		horizontalAdvanceWidth, leftSideBearing = ttFont["hmtx"].metrics[glyphName]
+		if not hasattr(glyph, 'xMin'):
+			glyph.recalcBounds(self)
+		leftSideX = glyph.xMin - leftSideBearing
+		rightSideX = leftSideX + horizontalAdvanceWidth
+		if "vmtx" in ttFont:
+			verticalAdvanceWidth, topSideBearing = ttFont["vmtx"].metrics[glyphName]
+			topSideY = topSideBearing + glyph.yMax
+		else:
+			# without vmtx, use ascent as vertical origin and UPEM as vertical advance
+			# like HarfBuzz does
+			verticalAdvanceWidth = ttFont["head"].unitsPerEm
+			try:
+				topSideY = ttFont["hhea"].ascent
+			except KeyError:
+				# sparse masters may not contain an hhea table; use the ascent
+				# of the default master as the vertical origin
+				assert defaultVerticalOrigin is not None
```

Comment:
caller is required to pass a `defaultVerticalOrigin` when getting phantom points from a sparse master font that may not contain an hhea table. I will change it so that the KeyError("hhea") is raised when defaultVerticalOrigin is not provided.

Reply:
not worth raising an error here, a warning shold be ok; if no vmtx and hhea tables are present and caller didn't bother passing a defaultVerticalOrigin, they probably don't care about the vertical phantom points. I can use the UPEM for both verticalAdvanceWidth and topSideY coord.